build-system/compile/internal-version.js

```diff
@@ -26,40 +26,75 @@ const argv = minimist(process.argv.slice(2), {
+  //   `master`'s commit history, then the current commit is <C>).
+  // - Find the commit time of <C> (<C>.time). Note that commit time might be
+  //   different from author time! e.g., commit time might be the time that a PR
+  //   was merged into `master`, or a commit was cherry-picked onto the brabnch;
+  //   author time is when the original author of the commit ran the
+  //   "git commit" command.
+  // - The version number is <C>.time.format("YYmmDDHHMMSS", "UTC") + <X> (the
+  //   pseudo-code assumes standard `.strftime` formatting).
+  //
+  // Examples:
+  // 1. The version number of a release built from the HEAD commit on `master`,
+  //    where that HEAD commit was committed on April 25, 2020 2:31:11 PM EDT
+  //    would be `2004251831110`.
+  //    - EDT is UTC-4, so the hour value changes from EDT's 14 to UTC's 18.
+  //    - The last digit is 0 as this commit is on `master`.
+  //
+  // 2. The version number of a release built from a local working branch (e.g.,
+  //    on a developer's workstation) that was split off from a `master` commit
+  //    from May 6, 2021 10:40:59 AM PDT and has multiple commits that exist
+  //    only on local working branch would be `2105061840590`.
+  //    - PDT is UTC-7, so the hour value changes from PST's 10 to UTC's 17.
+  //    - The last digit is 0 as this commit is on a branch that was split
+  //      from but does not have any commits since the split that were
+  //      cherry-picked from `master`.
+  //
+  // 3. For a release built from a local working branch that was split off from
+  //    a `master` commit from November 9, 2021 11:48:11 PM PST, and then:
+  //    - had one commit that was cherry-picked from `master`,
+  //    - followed by two commits that were created directly on the branch, the
+  //      last of which was commited on November 10, 2021 5:01:12 PM PST,
+  //    - followed by four commits that were cherry-picked from `master`,
+  //    then its version number would be `202111110101124`.
+  //    - The latest four commits are cherry-picks from `master`, and the one
+  //      before them is not, so our last digit is set to 4.
+  //    - PST is UTC-8, so the hour value changes from PST's 17 to UTC's 1, and
+  //      one day is added.
+  let numberOfCherryPicks = 0;
+  const commitCherriesInfo = gitCherryMaster().reverse();
+  for (const commitCherryInfo of commitCherriesInfo) {
+    if (!commitCherryInfo.isCherryPick) {
+      break;
     }
-    return `${lastCommitFormattedTime}${lastDigit}`;
+    numberOfCherryPicks++;
+  }
+  const lastCommitFormattedTime = gitCommitFormattedTime(
+    `HEAD~${numberOfCherryPicks}`
+  );
+
+  if (argv.strict_build && numberOfCherryPicks > 9) {
+    throw new Error(
+      `This branch has ${numberOfCherryPicks} cherry-picks. --strict_build caps the number of cherry-picks at 10. To build this branch use --version_override.`
+    );
   }
+  const lastDigit = Math.min(numberOfCherryPicks, 9);
```

Comment:
Nit: Seems unlikely to change until the next time we change versioning somehow, but I feel like `9` should be a constant `MAX_CHERRY_PICKS` or something

Reply:
Also, why do we want to allow >9 for non-strict builds?

Reply:
9 being the highest digit in our counting system, I feel like this is one of those exceptions to making it a const. Don't feel super strongly about this though.

The idea I had is that strict builds are what we create for releases - developers might have cherry-picks, regular commits, more cherry picks, etc - not failing their local builds seems like a good choice. If it all ends up being a PR it'll be squashed to a single commit at the end anyways and wouldn't interfere with the release process

Reply:
> I feel like 9 should be a constant `MAX_CHERRY_PICKS` or something

I agree that `9` here and on line 91 should be hoisted to a const at the top of this file.

> The idea I had is that strict builds are what we create for releases

Reopening this thread to further play out this scenario. I do see the rationale behind adding a flag to cap the cherry-pick count at 9. However, I don't see a graceful way to recover from the case where there are more than 9 cherry-picked commits in a given release branch (however unlikely it is in practice), and the strict cap is enforced.

- If we do enforce the cap of 9 CPs, we can't do any more releases from that branch and will have to cut a new one (correct?)
- If we don't enforce the cap, and there are two releases branches, one with 9 CPs and another with 10, they'll have the same version, which is also undesirable

I don't know a great solution to this with a single digit for CP status. In addition, I'm not super enthusiastic about adding yet another flag when `gulp dist` is invoked by the release automation, since maintenance and future flag changes require an elaborate 2 week process to avoid skew between release code and source code.

One alternative solution that comes to mind is to eliminate the two digits for `seconds` from CP version numbers, and use the last three digits to indicate CP status.

Pros as I see them:
- This will work the same for local development and releases
- No need for the additional flag
- CP release branches will be easy to identify and sort. (E.g. For the main branch of `2004161906340`, the first CP branch will be `2004161906000`, the second `2004161906001`, the third `2004161906002`, etc.)

Do you think this is feasible? Any drawbacks I'm not considering?

Reply:
After the discussion on the I2I, went with dropping the seconds